Duplicate ABI definitions of generic instantiations across modules

Can we use `#define`'s to avoid these?

```
include/SWRT/WindowsAppSDK.h:4310:8: error: redefinition of 'SWRT_WindowsFoundationCollections_IKeyValuePair_String_String'
4308 |
4309 | // Windows.Foundation.Collections.IKeyValuePair`2<String, String>
4310 | struct SWRT_WindowsFoundationCollections_IKeyValuePair_String_String
     |        `- error: redefinition of 'SWRT_WindowsFoundationCollections_IKeyValuePair_String_String'
4311 | {
4312 |     struct SWRT_WindowsFoundationCollections_IKeyValuePair_String_String_VirtualTable* VirtualTable;

include\SWRT\UWP.h:2087:8: note: previous definition is here
2085 |
2086 | // Windows.Foundation.Collections.IKeyValuePair`2<String, String>
2087 | struct SWRT_WindowsFoundationCollections_IKeyValuePair_String_String
     |        `- note: previous definition is here
2088 | {
2089 |     struct SWRT_WindowsFoundationCollections_IKeyValuePair_String_String_VirtualTable* VirtualTable;

<module-includes>:1:10: note: in file included from <module-includes>:1:
1 | #include "SWRT/WindowsAppSDK.h"
  |          `- note: in file included from <module-includes>:1:

--- a/Generator/Sources/CodeWriters/C/CSourceFileWriter.swift
+++ b/Generator/Sources/CodeWriters/C/CSourceFileWriter.swift
@@ -1,5 +1,5 @@
 public class CSourceFileWriter {
-    private let output: IndentedTextOutputStream
+    public let output: IndentedTextOutputStream
 
     public init(output: some TextOutputStream, pragmaOnce: Bool = true) {
         self.output = .init(inner: output)

--- a/Generator/Sources/CodeWriters/IndentedTextOutputStream.swift
+++ b/Generator/Sources/CodeWriters/IndentedTextOutputStream.swift
@@ -135,11 +135,12 @@ public class IndentedTextOutputStream: TextOutputStream {
     }
 
     public func writeIndentedBlock(
-        grouping: VerticalGrouping? = nil,
-        header: String? = nil,
-        footer: String? = nil,
-        endFooterLine: Bool = true,
-        body: () throws -> Void) rethrows {
+            grouping: VerticalGrouping? = nil,
+            header: String? = nil,
+            indentation: Int = 1,
+            footer: String? = nil,
+            endFooterLine: Bool = true,
+            body: () throws -> Void) rethrows {
 
         if let grouping {
             beginLine(grouping: grouping)
@@ -156,11 +157,11 @@ public class IndentedTextOutputStream: TextOutputStream {
         // Force the indented body to be grouped with the previous line
         self.lineGrouping = nil
 
-        indentLevel += 1
+        indentLevel += indentation
         try body()
         if case .end = lineState {}
         else { endLine() }
-        indentLevel -= 1
+        indentLevel -= indentation
 
         if let footer {
             // Force the footer to be grouped with the line above

--- a/Generator/Sources/ProjectionModel/CAbi/CAbi+writing.swift
+++ b/Generator/Sources/ProjectionModel/CAbi/CAbi+writing.swift
@@ -65,10 +65,22 @@ extension CAbi {
                 params: params)
         }
 
+        if !genericArgs.isEmpty {
+            // No module owns a generic type instantiation, so use #define guards to prevent multiple definitions.
+            let grouping = writer.output.allocateVerticalGrouping()
+            writer.output.writeFullLine(grouping: grouping, "#ifndef \(mangledName)")
+            writer.output.writeFullLine(grouping: grouping, "#define \(mangledName)", groupWithNext: true)
+        }
+
         decl.write(
             comment: try WinRTTypeName.from(type: boundType).description,
             forwardDeclared: true,
             to: writer)
+
+        if !genericArgs.isEmpty {
+            writer.output.endLine(groupWithNext: true)
+            writer.output.writeFullLine(grouping: .never, "#endif")
+        }
     }
 
     private static func appendCOMParams(name: String?, type: TypeNode, genericArgs: [TypeNode], isByRef: Bool, to comParams: inout [CParamDecl]) throws {

--- a/Generator/Sources/ProjectionModel/CAbi/CAbi+writing.swift
+++ b/Generator/Sources/ProjectionModel/CAbi/CAbi+writing.swift
@@ -69,7 +69,7 @@ extension CAbi {
             // No module owns a generic type instantiation, so use #define guards to prevent multiple definitions.
             let grouping = writer.output.allocateVerticalGrouping()
             writer.output.writeFullLine(grouping: grouping, "#ifndef \(mangledName)")
-            writer.output.writeFullLine(grouping: grouping, "#define \(mangledName)", groupWithNext: true)
+            writer.output.writeFullLine(grouping: grouping, "#define \(mangledName) \(mangledName)", groupWithNext: true)
         }
 
         decl.write(

--- a/InteropTests/WinRTComponent/IDL/IExposeObservableVector.idl
+++ b/InteropTests/WinRTComponent/IDL/IExposeObservableVector.idl
@@ -6,6 +6,6 @@ namespace WinRTComponent
     // This resulted in IObservableVectorBinding.*.Import to not implement the First() method.
     interface IExposeObservableVector
     {
-        IObservableVector<Int32> HereItIs { get; };
+        Windows.Foundation.Collections.IObservableVector<Int32> HereItIs { get; };
     }
 }

--- a/InteropTests/WinRTComponent/IDL/IMultiGenericInstantiation.idl
+++ b/InteropTests/WinRTComponent/IDL/IMultiGenericInstantiation.idl
@@ -1,0 +1,16 @@
+namespace WinRTComponent
+{
+    // Regression test for https://github.com/tristanlabelle/swift-winrt/issues/350
+    // where two ABI modules would contain the same generic instantiation,
+    // resulting in a multiply defined C symbol.
+    interface IMultiGenericInstantiation
+    {
+        // Force multiple instantiations of TypedEventHandler<IMemoryBufferReference, IInspectable>
+
+        // 1. In the UWP module via IMemoryBufferReference.Closed
+        Windows.Foundation.IMemoryBufferReference GetMemoryBufferReference();
+
+        // 2. In the WinRTComponement module via the method below
+        Windows.Foundation.TypedEventHandler<Windows.Foundation.IMemoryBufferReference, IInspectable> GetTypedEventHandler();
+    }
+}

--- a/InteropTests/WinRTComponent/IDL/WinRTComponent.idl
+++ b/InteropTests/WinRTComponent/IDL/WinRTComponent.idl
@@ -12,6 +12,7 @@
 #include "Int32Wrapper.idl"
 #include "InterfaceCasting.idl"
 #include "IExposeObservableVector.idl"
+#include "IMultiGenericInstantiation.idl"
 #include "IMultiOutParams.idl"
 #include "ManualAsyncOperation.idl"
 #include "MinimalTypes.idl"

